chain/tsresolver/tipset_resolver.go

```diff
@@ -0,0 +1,164 @@
+	EthBlockSelectorSafe      = "safe"
+	EthBlockSelectorFinalized = "finalized"
+)
+
+type TipSetLoader interface {
+	GetHeaviestTipSet() (ts *types.TipSet)
+	LoadTipSet(ctx context.Context, tsk types.TipSetKey) (*types.TipSet, error)
+	GetTipsetByHeight(ctx context.Context, h abi.ChainEpoch, anchor *types.TipSet, prev bool) (*types.TipSet, error)
+}
+
+type F3 interface {
+	GetLatestCert(ctx context.Context) (*certs.FinalityCertificate, error)
+	IsEnabled() bool
+	IsRunning() (bool, error)
+}
+
+type TipSetResolver interface {
+	ResolveEthBlockSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error)
+}
+
+type tipSetResolver struct {
+	loader TipSetLoader
+	f3     F3
+}
+
+var _ TipSetResolver = (*tipSetResolver)(nil)
+
+func NewTipSetResolver(loader TipSetLoader, f3 F3) TipSetResolver {
+	return &tipSetResolver{
+		loader: loader,
+		f3:     f3,
+	}
+}
+
+// ResolveEthBlockSelector resolves an Ethereum block selector string to a TipSet.
+//
+// The selector can be one of:
+//   - "pending": the chain head
+//   - "latest": the TipSet with the latest executed messages (head - 1)
+//   - "safe": the TipSet with messages executed at least eth.SafeEpochDelay (30) epochs ago or the
+//     latest F3 finalized TipSet
+//   - "finalized": the TipSet with messages executed at least policy.ChainFinality (900) epochs ago
+//     or the latest F3 finalized TipSet
+//   - a decimal block number: the TipSet at the given height
+//   - a 0x-prefixed hex block number: the TipSet at the given height
+//
+// If a specific block number is specified and `strict` is true, an error is returned if the block
+// number resolves to a null round, otherwise in the case of a null round the first non-null TipSet
+// immediately before the null round is returned.
+func (tsr *tipSetResolver) ResolveEthBlockSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error) {
+	switch selector {
+	case EthBlockSelectorEarliest:
+		return nil, fmt.Errorf(`block param "%s" is not supported`, EthBlockSelectorEarliest)
+
+	case EthBlockSelectorPending:
+		return tsr.loader.GetHeaviestTipSet(), nil
+
+	case EthBlockSelectorLatest:
+		// head - 1 because we're always one behind for Eth compatibility due to deferred execution
+		return tsr.loader.LoadTipSet(ctx, tsr.loader.GetHeaviestTipSet().Parents())
+
+	case EthBlockSelectorSafe, EthBlockSelectorFinalized:
+		defaultDelay := policy.ChainFinality
+		if selector == EthBlockSelectorSafe {
+			defaultDelay = ethtypes.SafeEpochDelay
+		}
+
+		head := tsr.loader.GetHeaviestTipSet()
+		latestHeight := head.Height() - 1 // always one behind for Eth compatibility due to deferred execution
+		defaultHeight := latestHeight - defaultDelay
+
+		if f3TipSet, err := tsr.getF3FinalizedTipSet(ctx); err != nil {
+			return nil, err
+		} else if f3TipSet != nil && f3TipSet.Height() > defaultHeight {
+			// return the parent of the finalized tipset (deferred execution, eth cares about t)
```

Comment:
Questions:
* we have already decremented the height by 1 in line 94; why get parent of tipset finalised by F3?
* I thought deferred execution in the context of eth only matters for `head`? I am not clear on this one, apologies.

Reply:
Thinking about the consequences of deferred execution:
If tipset T is finalized, then the parent state root is finalized, as are the messages executed in the next tipset. However, that execution state, in theory, is not finalized.

So, both interpretations are valid.

Reply:
1. I think there is a calculation error here with comparing to a value that's already decremented
2. The tipset-1 thing is how we've baked in this behaviour to the existing Eth APIs, but I've been wondering whether F3 might be a chance to revisit current practice for `safe` and `finalized`. I'm going to have to go back and do a quick audit of the APIs that consume this to figure out what guarantees we need and whether the fixed list of messages is enough, or whether we need them to have been executed. Events are important for many of these APIs and they are attached to the next tipset, which is why head-1 is an important case.

Reply:
I'm going to defer (har har) dealing with the -1 question for `"safe"` and `"finalized"` for a possible follow-up PR, since this is only replicating existing behaviour. I've dealt with the other concern (the decrement was fine, but I've tried to make the code more clear and added a comment)